### PR TITLE
Configuration flags for cluster autoscaler

### PR DESCRIPTION
**Pull request by mwielgus**
https://github.com/kubernetes/kubernetes/issues/24404

### CODE REVIEW

File: cluster-autoscaler/config/migconfig.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	MaxSize int
+	Url     url.URL
+}
+
+type MigConfigFlag []MigConfig
+
+func (this *MigConfigFlag) String() string {
+	var b bytes.Buffer
+	b.WriteString("[")
+	for i, npc := range *this {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: What is `npc`? Please use more descriptive names.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 2]
```diff
@@ -16,5 +16,18 @@ limitations under the License.
...
+	"fmt"
+	"k8s.io/contrib/cluster-autoscaler/config"
+)
+
+var (
+	migConfig config.MigConfigFlag
+)
+
 func main() {
+	flag.Var(&migConfig, "nodes", "sets min,max size and url of a MIG to be controlled by Cluster Autoscaler")
```
Comment (fgrzadkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: Please explain that it's possible to use this flag multiple times to pass control more MIGs.

Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Done.

File: cluster-autoscaler/config/migconfig.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+type MigConfig struct {
+	MinSize int
+	MaxSize int
+	Url     url.URL
+}
+
+type MigConfigFlag []MigConfig
+
+func (this *MigConfigFlag) String() string {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Can we just use combination of `fmt.Sprintf` and `strings.Join`? It'll be much more readable.

File: cluster-autoscaler/config/migconfig.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Fixed.

File: cluster-autoscaler/config/migconfig.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		b.WriteString(fmt.Sprintf("%d:%d:%s", npc.MinSize, npc.MaxSize, npc.Url.String()))
+	}
+	b.WriteString("]")
+	return b.String()
+}
+
+func (this *MigConfigFlag) Set(value string) error {
+	tokens := strings.SplitN(value, ":", 3)
+	if len(tokens) != 3 {
+		return fmt.Errorf("wrong nodes configuration: %s", value)
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Please add a correct format.

File: cluster-autoscaler/config/migconfig.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	ncf := MigConfig{}
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Please don't use names such as `ncf`. I have no idea what this means. It makes this code much harder to read.

File: cluster-autoscaler/config/migconfig.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	if size, err := strconv.Atoi(tokens[0]); err == nil {
+		ncf.MinSize = size
+	} else {
+		return fmt.Errorf("failed to set min size: %s", tokens[0])
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: "failed to set min size: %s; expected integer"

File: cluster-autoscaler/config/migconfig.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (this *MigConfigFlag) Set(value string) error {
+	tokens := strings.SplitN(value, ":", 3)
+	if len(tokens) != 3 {
+		return fmt.Errorf("wrong nodes configuration: %s", value)
+	}
+	ncf := MigConfig{}
+	if size, err := strconv.Atoi(tokens[0]); err == nil {
+		ncf.MinSize = size
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Do we accept negative values? Same applies to max.

File: cluster-autoscaler/config/migconfig.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	} else {
+		return fmt.Errorf("failed to set min size: %s", tokens[0])
+	}
+
+	if size, err := strconv.Atoi(tokens[1]); err == nil {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: We should check if `min<max`. Do we accept `min=max`?

File: cluster-autoscaler/config/migconfig.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	} else {
+		return fmt.Errorf("failed to set max size: %s", tokens[1])
+	}
+
+	if url, err := url.Parse(tokens[2]); err == nil {
+		ncf.Url = *url
+	} else {
+		return fmt.Errorf("failed to parse mig url: %s", tokens[2])
+	}
+	if len(ncf.Url.String()) < 5 {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Why `5`?

File: cluster-autoscaler/config/migconfig_test.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+package config
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSet(t *testing.T) {
+	mcf := MigConfigFlag{}
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig_test.go]: Please rename - `config`, `configFlag` or `migConfigFlag`?

File: cluster-autoscaler/config/migconfig_test.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	assert.Error(t, mcf.Set("a"))
+	assert.Error(t, mcf.Set("a:b:c"))
+	assert.Error(t, mcf.Set("1:2:x"))
+	assert.Error(t, mcf.Set("1:2:"))
+	assert.NoError(t, mcf.Set("111:222:http://a.b.c"))
+	assert.Contains(t, mcf.String(), "111")
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig_test.go]: Can you explicitly check `min`, `max` and `url`?`

File: cluster-autoscaler/config/migconfig.go (hunk 6 above)
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: or "positive integer" or something like this.

File: cluster-autoscaler/config/migconfig.go (hunk 7 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Done.

File: cluster-autoscaler/config/migconfig.go (hunk 8 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Yes, we do. Done.

File: cluster-autoscaler/config/migconfig.go (hunk 9 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: url.Parse accepts empty strings and strings that do not look like urls. anything less than couple charecters is most likely not a valid mig.

File: cluster-autoscaler/config/migconfig.go (hunk 6 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Done.

File: cluster-autoscaler/config/migconfig.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Done.

File: cluster-autoscaler/config/migconfig.go (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: Done.

File: cluster-autoscaler/config/migconfig_test.go (hunk 11 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig_test.go]: Done.

File: cluster-autoscaler/config/migconfig.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	} else {
+		return fmt.Errorf("failed to set max size: %s, expected integer", tokens[1])
+	}
+
+	if url, err := url.Parse(tokens[2]); err == nil {
+		migconfig.Url = *url
+	} else {
+		return fmt.Errorf("failed to parse mig url: %s", tokens[2])
+	}
+	if len(migconfig.Url.String()) < 5 {
```
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Please add a comment why `5`.

File: cluster-autoscaler/config/migconfig.go (hunk 4 above)
Comment (fgrzadkowski) [inline cluster-autoscaler/config/migconfig.go]: Not fixed yet.

File: cluster-autoscaler/config/migconfig_test.go (hunk 10 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig_test.go]: done

File: cluster-autoscaler/config/migconfig.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/config/migconfig.go]: In flag descriptor.

**Pull request merged**